jax/_src/numpy/lax_numpy.py

```diff
@@ -2262,17 +2262,43 @@ def _convert_to_array_if_dtype_fails(x: ArrayLike) -> ArrayLike:
 In particular, the details of float-to-int and int-to-float casts are
 implementation dependent.
 """)
-def astype(x: ArrayLike, dtype: DTypeLike | None, /, *, copy: bool = True) -> Array:
+def astype(x: ArrayLike, dtype: DTypeLike | None,
+           /, *, copy: bool = True,
+           device: xc.Device | Sharding | None = None) -> Array:
   util.check_arraylike("astype", x)
   x_arr = asarray(x)
-  del copy  # unused in JAX
   if dtype is None:
     dtype = dtypes.canonicalize_dtype(float_)
   dtypes.check_user_dtype_supported(dtype, "astype")
+  if (
+    issubdtype(x_arr.dtype, complexfloating)
+    and dtypes.isdtype(dtype, ("integral", "real floating"))
+  ):
+    warnings.warn(
+      "Casting from complex to real dtypes will soon raise a ValueError. "
+      "Please first use jnp.real or jnp.imag to take the real/imaginary "
+      "component of your input.",
+      DeprecationWarning, stacklevel=2
+    )
+
+  out = x_arr
   # convert_element_type(complex, bool) has the wrong semantics.
   if np.dtype(dtype) == bool and issubdtype(x_arr.dtype, complexfloating):
```

Comment:
We should put both these `issubdtype(x_arr.dtype, complexfloating)` checks in a single block